### ISSUE TITLE
Expose TCState in the reflection API

### ISSUE THREAD

**Issue by jespercockx** [agda/agda#3346]
Currently, the built-in way to implement tactics with backtracking in the `TC` monad is by using `throwError` and `catchError`. However, this catches **all** errors which is often not what you want and enforces a bad programming style.

So I'd rather implement my own backtracking mechanism in Agda, but currently that isn't really possible because there is no way to 'roll back' the TC state (metavariables and constraints) without `catchTC`. If we had some build-in functions to directly get and put the TC state, then it'd be possible to implement backtracking in Agda on top of that.

Of course, if there is another less low-level primitive we could add to the reflection API that can be used for the same purpose, I'd be happy with that too.

**Comment by nad** [comment]
You don't intend to allow direct modification of components of the state, right?

**Comment by jespercockx** [comment]
I just want to be able to save the TC state and restore it, I don't need any direct access to its fields (as that would indeed be quite unsafe).

**Comment by jespercockx** [comment]
Another solution to my problem would be to have some primitive to run a given `TC` action in 'speculative mode', and decide based on the result whether to keep the new TC state or roll back to the old one.

One way you can get something close to this with the current infrastructure is by using `catchTC`:
```agda
open import Prelude
open import Builtin.Reflection

-- Runs the given action speculatively,
-- committing to the new state if the result
-- is true or rolling back if the result is false
runSpeculative : TC Bool → TC ⊤
runSpeculative m = catchTC
  (do
     x ← m
     if x then return _ else typeError [])
  (return _)
```
This isn't quite enough to implement 'true' backtracking for two reasons:
1. `catchTC` catches all errors thrown by `m`, while I really just want to catch the one error I threw myself.
2. It doesn't allow us to return any result from the speculative operation.

Ideally, we'd want to run a speculative operation and do something with the result, but if the result is later rejected (i.e. there's a type error thrown by the tactic), then we want to backtrack **to the point of the speculative computation** and take the alternative branch there.

I feel confident that I could implement a backtracking algorithm like this on top of the reflection API if I had some way to save/reset the state, but the current primitives just seem to be too limited for this purpose.

(One other thing I tried was to use continuation-passing so the entire continuation of the speculative computation would be 'captured' by the `catchTC`. But this lead to other problems that I'm currently unable to reconstruct. I could try again to make it work, but it seemed much harder to reason about in general than the more first-order approach I'm currently trying).

ps: I also looked into the `commitTC` primitive, but it seems to me it's more specialized and doesn't help for the problem here.

**Comment by jespercockx** [comment]
Adding a new primitive type for `TCState` is problematic because we'd need to make `TCState` part of the syntax which leads to the unholy mother of all import loops. Instead, I'll try to add a new primitive
```
  runSpeculativeTC : TC (A × Bool) -> TC A
```
with the following semantics:

- If running `m` returns `x , true`, then the effect of `runSpeculative m` is the same as running `fst <$> m` (i.e. the state is kept)
- If running `m` returns `x , false`, then the effect of `runSpeculative m` is the same as running `return x` (i.e. the state is reset to its previous value)

This seems like it should enable all the backtracking magic without having to add a new primitive type.

**Comment by jespercockx** [comment]
I added the new primitive (on my own branch) and tested it on my WIP tactics library, it seems to work as expected:
```agda
  runTac' (chooseTac tac₁ tac₂) goal cont = do
      x ← runSpeculative $ do
        just subgoals ← runTac' tac₁ goal cont
          where nothing → return $ nothing , false
        return $ just subgoals , true
      case x of λ where
        nothing         → runTac' tac₂ goal cont
        (just subgoals) → return $ just subgoals
```
https://github.com/jespercockx/ataca/blob/2c8411b2e839ef6ea71ec2bbf7c8d2dc82da5a79/src/Tac.agda#L38-L45

As the feature freeze is still in place I'll wait with pushing until after the release.

**Comment by jespercockx** [comment]
This was released in 2.6.0, I just forgot to close it.
